lib/rules/no-dupe-else-if.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * @fileoverview Rule to disallow duplicate conditions in if-else-if chains
+ * @author Milos Djermanovic
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const astUtils = require("./utils/ast-utils");
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        type: "problem",
+
+        docs: {
+            description: "disallow duplicate conditions in if-else-if chains",
+            category: "Possible Errors",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/no-dupe-else-if"
+        },
+
+        schema: [],
+
+        messages: {
+            unexpected: "Duplicate condition in if-else-if chain."
+        }
+    },
+
+    create(context) {
+        const sourceCode = context.getSourceCode();
+
+        return {
+            IfStatement(node) {
+                let current = node;
+
+                while (current.parent && current.parent.type === "IfStatement" && current.parent.alternate === current) {
+                    current = current.parent;
+
+                    if (astUtils.equalTokens(node.test, current.test, sourceCode)) {
```

Comment:
I think it's a very good idea.

We could even compare with the union of `||` conditions from all parents in the chain?

For example:

```js
if (a) {
    // ...
} else if (b) {
    // ...
} else if (a || b) { // duplicated condition: { a, b } is a subset of { a, b }
    // ...
}
```

That would also catch things like this:

```js
if (a || b) {
    // ...
} else if (c || d) {
    // ...
} else if (a || d) { // duplicated condition: { a, d } is a subset of { a, b, c, d }
    // ...
}
```

Reply:
Nice!

Reply:
Is this really necessary? This seems like it would add some complexity and go beyond the original proposal.

Reply:
Looks like there are  3 options at the moment:

1. To check for equal ancestors only, as originally proposed in #12469
2. To check for a subset of an ancestor as in [this comment](https://github.com/eslint/eslint/pull/12504#discussion_r340002860)
3. To check for a subset of the union of all ancestors as in [this comment](https://github.com/eslint/eslint/pull/12504#discussion_r340046553)

I'm fine with all three :+1: 

Options 2. and 3. look very nice, but only 1. is accepted, so I don't how to proceed :confused:  

It doesn't look too complicated to do 2. or 3. Though, compared to the simple 1. it would be certainly much more code. Not sure how to weight how useful these enhancements would be.

Reply:
I don't have a strong opinion either way at this point, but I'll point out that it's perfectly okay to implement option 1 first, and then see if we want to enhance the rule later based on user feedback. I don't strongly oppose option 2 or 3, I just want to make sure the benefit is worth the maintenance cost.

Reply:
Here's an implementation of option 3, for comparison: [no-dupe-else-if.js](https://github.com/mdjermanovic/eslint/blob/55b3257b9fb111f6c970be993eebcae4b3b34749/lib/rules/no-dupe-else-if.js)

(it wasn't much work, I won't mind if we keep option 1)

Reply:
Great. That looks simple enough. I believe that option 3. fits the mind of this rule because it reports only unreachable `if` blocks by duplicated condition

Reply:
I can copy that to this PR + add more tests and docs. Also, the message should be probably changed, since the reported node might not be the exact 'duplicate' of a previous condition.

But better to check first if we have a consensus to implement option 3 from the start?

Reply:
I don't have any complexity concerns anymore based on the implementation shared earlier.

I don't understand why this only applies to `||` operators only, though. Shouldn't we look at `&&` as well?

Reply:
I like this addition! Agreed with @platinumazure. It does seem like it could be useful to also check for and warn on:

```js
if (a && b) {} 
else if (b && a) {}
```

Reply:
:+1: to account for `&&` as well. 

The logic for `&&` in a chain is a bit different, so option 3 wouldn't apply well. For instance, the following is valid code:

```js
if (a && b) {
    // ...
} else if (a) {
    // ...
} else if (b) {
    // ... 
}

if (a && b) {
    // ...
} else if (b && c) {
    // ...
} else if (a && c) {
    // ...
}
```

On the other hand, 'inverted' option 2 should work quite well for `&&` and is easy to implement. If the condition split by `&&` is a superset of one of the previous conditions, then it's a 'duplicate'. For example, all these are invalid:

```js
if (a) {
    // ...
} else if (a && b)

if (a && b) {
    // ...
} else if (b && a)

if (a && b) {
    // ...
} else if (a && b && c)
```

In addition, it also looks easy to split the original condition by `&&` and do the `||` check with each element of the resulting array, which would catch cases that might not even look like a bug at first glance:

```js
if (a || b) {
    // ...
} else if (b && c) { // perhaps a not-so-obvious bug
    // ...  
}

if (a) {
    // ...
} else if (b) {
    // ...
} else if ((a || b) && c) {
   // ...
}
```

That shouldn't even add new lines, just `.map` and `.some` to the existing lines in [this draft](https://github.com/mdjermanovic/eslint/blob/55b3257b9fb111f6c970be993eebcae4b3b34749/lib/rules/no-dupe-else-if.js)

Reply:
This is done now, the rule should be able to catch some relatively complex errors ('more complex errors with || and &&' section in `invalid` tests).

Examples from the documentation:

```js
/*eslint no-dupe-else-if: "error"*/

if (a || b) {
    foo();
} else if (a) {
    bar();
}

if (a) {
    foo();
} else if (b) {
    bar();
} else if (a || b) {
    baz();
}

if (a) {
    foo();
} else if (a && b) {
    bar();
}

if (a && b) {
    foo();
} else if (a && b && c) {
    bar();
}

if (a || b) {
    foo();
} else if (b && c) {
    bar();
}

if (a) {
    foo();
} else if (b && c) {
    bar();
} else if (d && (c && e && b || a)) {
    baz();
}
```

Reply:
Tried the rule on ESLint codebase, reports only 1 bug at [indent-legacy.js#L779](https://github.com/eslint/eslint/blob/586855060afb3201f4752be8820dc85703b523a6/lib/rules/indent-legacy.js#L779) and it's indeed a bug because of `!parentVarNode` which is used in the previous condition with `||`.